lib/libutils/ext/include/fault_mitigation.h

```diff
@@ -0,0 +1,687 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2022, Linaro Limited
+ */
+#ifndef __FAULT_MITIGATION_H
+#define __FAULT_MITIGATION_H
+
+#include <assert.h>
+#include <config.h>
+#include <string.h>
+#include <util.h>
+
+#ifdef __KERNEL__
+#include <kernel/panic.h>
+#include <kernel/thread.h>
+#else
+#include <tee_api.h>
+#endif
+
+/*
+ * Fault migitigation helpers to make successful Hardware Fault Attacks
+ * harder to achieve. The paper [1] by Riscure gives background to the
+ * problem.
+ *
+ * These helpers aim to make it hard for a single glitch attack to succeed
+ * while the protected function or one of the ftmn_*() functions are
+ * executed.
+ *
+ * To have something to work with we assume that a single glitch may affect
+ * a few instructions in sequence to do nothing or to corrupt the content
+ * of a few registers.
+ *
+ * Using the terminology from [1] we are implementing the following patterns:
+ * 3 FAULT.VALUE.CHECK
+ * 5 FAULT.DECISION.CHECK
+ * 9 FAULT.FLOW.CONTROL
+ *
+ * Additionally are the following patterns also racknowledged with a few
+ * comments:
+ * 1. FAULT.CONSTANT.CODING
+ *	Zero is normally a success code in OP-TEE so special functions are
+ *	added to record anything but a zero result.
+ * 8. FAULT.NESTED.CHECK
+ *	The linked calls performed by for instance FTMN_CALL_FUNC() addresses
+ *	this by relying on the called function to update a state in
+ *	struct ftmn_func_arg which is checked when the function has returned.
+ * 11. FAULT.PENALTY
+ *	This is implicit since we're normally trying to protect things post
+ *	boot and booting takes quite some time.
+ *
+ * [1] https://www.riscure.com/uploads/2020/05/Riscure_Whitepaper_Fault_Mitigation_Patterns_final.pdf
```

Comment:
@jenswi-linaro the link is not visible.

Reply:
There is a copy at https://web.archive.org/web/20220616035354/https://www.riscure.com/uploads/2020/05/Riscure_Whitepaper_Fault_Mitigation_Patterns_final.pdf perhaps we could use that link?

Reply:
Thanks, I'll update.